crates/sui-protocol-config/src/lib.rs

```diff
@@ -511,4 +530,43 @@ impl ProtocolConfig {
             _ => panic!("unsupported version {:?}", version),
         }
     }
+
+    /// Override one or more settings in the config, for testing.
+    /// This must be called at the beginning of the test, before get_for_(min|max)_version is
+    /// called, since those functions cache their return value.
+    pub fn apply_overrides_for_testing(
+        override_fn: impl Fn(ProtocolVersion, Self) -> Self + Send + 'static,
+    ) -> OverrideGuard {
+        CONFIG_OVERRIDE.with(|ovr| {
+            let mut cur = ovr.lock().unwrap();
+            assert!(cur.is_none(), "config override already present");
+            *cur = Some(Box::new(override_fn));
+            OverrideGuard
+        })
+    }
+}
+
+// Setters for tests
+impl ProtocolConfig {
+    pub fn set_max_function_definitions_for_testing(&mut self, m: usize) {
+        self.max_function_definitions = Some(m)
+    }
+}
+
+type OverrideFn = dyn Fn(ProtocolVersion, ProtocolConfig) -> ProtocolConfig + Send;
+
+thread_local! {
+    static CONFIG_OVERRIDE: Mutex<Option<Box<OverrideFn>>> = Mutex::new(None);
```

Comment:
what is the `Mutex` for here.

Reply:
well I could have used RefCell but I was too lazy to import something else 🤣

Reply:
I guess a `RefCell` would have been a bit more clear but not a big deal...